src/actions.ts

```diff
@@ -0,0 +1,449 @@
+		return res;
+	},
+	async json() {
+		if (!Dispatcher.isJSON(this.request)) {
+			throw new ActionError("/api/json must use application/json requests");
+		}
+		let json;
+		try {
+			json = await Dispatcher.parseSentRequest(this.request);
+		} catch {
+			return [{actionerror: 'Malformed JSON sent.'}];
+		}
+		if (!json || !Array.isArray(json)) {
+			throw new ActionError(`Malformed JSON (send a JSON array in the 'json' property).`);
+		}
+
+		let serverid, servertoken;
+		for (const req of json) {
+			if (!serverid) serverid = req.serverid;
+			if (!servertoken) servertoken = req.servertoken;
+		}
+		const server = Dispatcher.servers[toID(serverid)];
+		if (json.length > 20) {
+			if (!server || server.token && server.token !== md5(servertoken)) {
+				throw new ActionError(`Only registered servers can send >20 requests at once.`);
+			}
+		}
+
+		const results = [];
+		for (const request of json) {
+			if (request.actionerror) continue;
+			if (!request.act) {
+				results.push({actionerror: 'Must send a request type.'});
+				continue;
+			}
+			const dispatcher = new Dispatcher(this.request, this.response, {
+				body: request,
+				act: request.act,
+			});
+			try {
+				const result = await dispatcher.executeActions();
+				results.push(result);
+			} catch (e) {
+				if (e instanceof ActionError) {
+					results.push({actionerror: e.message});
+					continue;
+				}
+				throw e;
+			}
+		}
+		return results;
+	},
+	async prepreplay(params) {
+		const server = await this.getServer(true);
+		if (!server) {
+			return {errorip: this.getIp()};
+		}
+		const extractedFormatId = /^([a-z0-9]+)-[0-9]+$/.exec((`${params.id}`));
+		const formatId = /^([a-z0-9]+)$/.exec((`${params.format}`));
+		if (
+			// the server must be registered
+			!server ||
+			// the server must send all the required values
+			!params.id ||
+			!params.format ||
+			!params.loghash ||
+			!params.p1 ||
+			!params.p2 ||
+			// player usernames cannot be longer than 18 characters
+			(params.p1.length > 18) ||
+			(params.p2.length > 18) ||
+			// the battle ID must be valid
+			!extractedFormatId ||
+			// the format ID must be valid
+			!formatId ||
+			// the format from the battle ID must match the format ID
+			(formatId[1] !== extractedFormatId[1])
+		) {
+			return 0;
+		}
+
+		if (server.id !== Config.mainserver) {
+			params.id = server.id + '-' + params.id;
+		}
+		params.serverid = server.id;
+
+		const result = await Replays.prep(params);
+
+		this.setPrefix(''); // No need for prefix since only usable by server.
+		return result;
+	},
+	uploadreplay(params) {
+		this.setHeader('Content-Type', 'text/plain; charset=utf-8');
+		return Replays.upload(params, this);
+	},
+	async invalidatecss() {
+		const server = await this.getServer(true);
+		if (!server) {
+			return {errorip: this.getIp()};
+		}
+		// No need to sanitise server['id'] because it should be safe already.
+		const cssfile = `${__dirname}/../config/customcss/${server['id']}.css`;
```

Comment:
Don't force the custom CSS to be in the directory tree of the code, take the directory as an environment variable or put it in the config file

Reply:
OK. Did this.